### PR TITLE
Android free rotation.

### PR DESCRIPTION
**Pull request by hysw**
- Discard VK_SUBOPTIMAL_KHR error.
- Handling content rect changed events.

### CODE REVIEW

**Comment by chaoticbob** [comment]
I'd like to see a fuller explanation of why this change his necessary. Based on the comments, there seems to be some confusion on event ordering on the desktop. The window resize event should always occur before render is called so neither Acquire or Present returns an error or a warning. Are the events on Android out of order where SUBOPTIMAL is actually showing up?

**Review by chaoticbob — CHANGES_REQUESTED** [review]
Needs clarification on why this is necessary.

**Comment by hysw** [comment]
> I'd like to see a fuller explanation of why this change his necessary. Based on the comments, there seems to be some confusion on event ordering on the desktop. The window resize event should always occur before render is called so neither Acquire or Present returns an error or a warning.

On Windows, it will probably not going happen.
On Linux it's currently possible to have `VK_ERROR_OUT_OF_DATE_KHR` on Present on Intel/X11/KWin.
I've yet to test all combination of {AMD,Nvidia,Intel}x{X11,Wayland}x{mutter,kwin,...} to see if they toss out `OUT_OF_DATE` or `SUBOPTIMAL`, and driver/x11/wayland might change that during any update.

Also technically, `VK_SUBOPTIMAL_KHR` is a "Success Codes" per Vulkan spec.

> Are the events on Android out of order where SUBOPTIMAL is actually showing up?

On Android, unless it's portrait and correct side up, we will always get `VK_SUBOPTIMAL_KHR`. It's more of a compositor asking for application to do pre-transform to match the hardware buffer so it does not have to. See [Arm's Documentation](https://github.com/ARM-software/vulkan_best_practice_for_mobile_developers/blob/master/samples/performance/surface_rotation/surface_rotation_tutorial.md#swapchain-re-creation)

E.g. on Landscape, we need to set `preTransform` to `VK_SURFACE_TRANSFORM_ROTATE_{90,270}_BIT_KHR`, swap swapchain height and width, and render side way.

**Comment by chaoticbob** [comment]
> On Android, unless it's portrait and correct side up, we will always get `VK_SUBOPTIMAL_KHR`. It's more of a compositor asking for application to do pre-transform to match the hardware buffer so it does not have to. See [Arm's Documentation](https://github.com/ARM-software/vulkan_best_practice_for_mobile_developers/blob/master/samples/performance/surface_rotation/surface_rotation_tutorial.md#swapchain-re-creation)
> 
> E.g. on Landscape, we need to set `preTransform` to `VK_SURFACE_TRANSFORM_ROTATE_{90,270}_BIT_KHR`, swap swapchain height and width, and render side way.

I'm not seeing where in the documentation where it says that the swapchain is always SUBOPTIMAL unless it's portrait and correctg side up? All I see it saying is that the swapchain's transform bit needs to mwatch what the surface is asking for so the compositor is happy. Am I missing something?

**Comment by hysw** [comment]
> I'm not seeing where in the documentation where it says that the swapchain is always SUBOPTIMAL unless it's portrait and correctg side up? All I see it saying is that the swapchain's transform bit needs to mwatch what the surface is asking for so the compositor is happy. Am I missing something?

https://github.com/google/bigwheels/blob/1bc7e11083ee7ffb5794f147087e4fb96c6e4ae5/src/ppx/grfx/vk/vk_swapchain.cpp#L409

The assumption of `VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR` is everywhere, so I'm not changing `preTransform`, and therefore the compositor is not happy about that.

We could fix the identity transform assumption, it will take quite a bit time fixing all the matrices, and/or re-compute and re-uploading vertex data for shaders that does not take transform matrix.

**Comment by chaoticbob** [comment]
> > I'm not seeing where in the documentation where it says that the swapchain is always SUBOPTIMAL unless it's portrait and correctg side up? All I see it saying is that the swapchain's transform bit needs to mwatch what the surface is asking for so the compositor is happy. Am I missing something?
> 
> https://github.com/google/bigwheels/blob/1bc7e11083ee7ffb5794f147087e4fb96c6e4ae5/src/ppx/grfx/vk/vk_swapchain.cpp#L409
> 
> The assumption of `VK_SURFACE_TRANSFORM_IDENTITY_BIT_KHR` is everywhere, so I'm not changing `preTransform`, and therefore the compositor is not happy about that.

What happens if you do something like this:
```
vkci.preTransform = ToApi(pCreateInfo->pSurface)->GetCapabilities().currentTransform;
```

> We could fix the identity transform assumption, it will take quite a bit time fixing all the matrices, and/or re-compute and re-uploading vertex data for shaders that does not take transform matrix.

The surface transform is only for the surface, it shouldn't affect the content. The disparity in the viewport/scissor for device rotation needs to be addressed differently.

**Comment by hysw** [comment]
>The surface transform is only for the surface, it shouldn't affect the content. The disparity in the viewport/scissor for device rotation needs to be addressed differently.

We basically get a vertical/portrait swapchain if we specify the preTransform equal to the surface capability. (this is checked on a Pixel 5)

**Comment by chaoticbob** [comment]
> > The surface transform is only for the surface, it shouldn't affect the content. The disparity in the viewport/scissor for device rotation needs to be addressed differently.
> 
> We basically get a vertical/portrait swapchain if we specify the preTransform equal to the surface capability. (this is checked on a Pixel 5)

It looks like vkCreateSwapchain isn't smart about the width height. The Arm documentation says the width and height needs to be swapped, probably someting like this is needed:

```
if (transform == VK_SURFACE_TRANSFORM_ROTATE_90_BIT_KHR || transform == VK_SURFACE_TRANSFORM_ROTATE_270_BIT_KHR)
{
	// Pre-rotation: always use native orientation i.e. if rotated, use width and height of identity transform
	std::swap(width, height);
}
```

Will probably need to handle the correct event on the app side as well to set teh window width/height.

**Comment by hysw** [comment]
>It looks like vkCreateSwapchain isn't smart about the width height. The Arm documentation says the width and height needs to be swapped, probably someting like this is needed:
> ```
> if (transform == VK_SURFACE_TRANSFORM_ROTATE_90_BIT_KHR || transform == VK_SURFACE_TRANSFORM_ROTATE_270_BIT_KHR)
> {
> 	// Pre-rotation: always use native orientation i.e. if rotated, use width and height of identity transform
> 	std::swap(width, height);
> }
> ```
> Will probably need to handle the correct event on the app side as well to set teh window width/height.

It's that we are creating a vertical swapchain when the surface is horizontal.

So now we have a swapchain/framebuffer width/height that is vertical, and content width/height that is horizontal. The content is also rendered side way respect to the swapchain/framebuffer orientation. I'm not even sure if window width/height even mean anything at this point.

Also there's the problem of Android virtual keyboard, and we get to deal with `android_app->contentRect`.

**Comment by chaoticbob** [comment]
I looked into this a bit more in depth. 

*TL;DR;* Most of your changes are fine. The change to ignore SUBOPTIMAL is not desirable, warnings shouldn't be blanked suppressed without a reason. I've made some minor changes to your CL, you can see them here: https://github.com/google/bigwheels/tree/android-rotation

Regarding ignoring SUBOPTIMAL - this behavior already existed in `grfx::vk::Swapchain::AcquireNextImageInternal`:
https://github.com/google/bigwheels/blob/main/src/ppx/grfx/vk/vk_swapchain.cpp#L562-L564.
It only logs the warning, it doesn't assert. There was an incorrect assumption in `grfx::vk::Swapchain::PresentInternal` that assumed that `vkres != VK_SUCCESS` is an error. The changes I made just copies the behavior in *acquire* for *present*.

I also added an `#ifdef` around the warnings so it doesn't flood logcat on Android. 

The current behavior in the changes I made on top of your is what we want to start out with. It's not the most efficient, but it's good enough for us to work with. There's obviously some app bugs that need to correct their aspects. We also don't have anyway of telling the applications that they need to prepend their view matrices with an orientation adjustment. Later CLs will address these issues.

My recommendation is to combine my change to yours and make whatever adjustments you need to.

**Comment by hysw** [comment]
> Without rotation between background/foreground, the app is fine, but doing a rotation while backgrounded seems to break it. Maybe some path we don't handle in native yet?

It's not the rotation causing problem, app suspension is not supported yet.

File: src/android/AndroidManifest.xml [hunk 1]
```diff
@@ -11,6 +11,8 @@
       tools:targetApi="31">
     <activity
         android:name="com.google.bigwheels.MainActivity"
+        android:configChanges="orientation|screenSize"
```
[the remark this answers is not in the document]
Comment (hysw) [reply src/android/AndroidManifest.xml:14]: No, OpenXR has it's own thing, screen size don't change, and orientation is a quaternion.
Well, at least for HMD/STEREO display.

**Comment by chaoticbob** [comment]
> > Without rotation between background/foreground, the app is fine, but doing a rotation while backgrounded seems to break it. Maybe some path we don't handle in native yet?
> 
> It's not the rotation causing problem, app suspension is not supported yet.

Lets deal with the suspend / wake in another PR. It's pretty involved IIRC.

**Comment by chaoticbob** [comment]
> Agreeing with @chaoticbob feedback. From testing, it does solve rotation when the application is running, but seems like the following is not handled:
> 
> * start app in portrait

Can you provide some detail on this? I just tested and the cube sample is able to startup in both portrait and landscape mode. Unless I'm misunderstanding that what portrait means here.

**Comment by hysw** [comment]
> Can you provide some detail on this?

I'm assuming @Keenuts 's bullet point is actually ordered list of step to reproduce a breakage. Which the app stop responding after resume from background. This is the suspend/resume issue, which will not be addressed in this PR.

**Comment by hysw** [comment]
> My recommendation is to combine my change to yours and make whatever adjustments you need to.

Did a few adjustment on top:
* Log VK_SUBOPTIMAL_KHR once on Android.
* After reading some source code, it seems contentRect is complicated, so ignore it for now, `WINDOW_RESIZED` works for portrait/lamdscape mode.
* Also check Surface.cap.currentExtent, because apparently ARM/Khronos is using that for swapchain size.

Also note that ARM best practice docs is actually in [Khronos Vulkan-Samples](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/performance/surface_rotation) now

File: include/ppx/log.h [hunk 2]
```diff
@@ -132,6 +132,18 @@ class Log
         ppx::Log::Get()->Unlock();                   \
     }
 
+#define PPX_LOG_WARN_ONCE(MSG)                         \
+    if (ppx::Log::IsActive()) {                        \
+        ppx::Log::Get()->Lock();                       \
+        static bool ppxLogWarnOnce = false;            \
```
[the remark this answers is not in the document]
Comment (chaoticbob) [reply include/ppx/log.h:138]: NVM, verified it's scope to the enclosing function.

**Comment by chaoticbob** [comment]
> > My recommendation is to combine my change to yours and make whatever adjustments you need to.
> 
> Did a few adjustment on top:
> 
> * Log VK_SUBOPTIMAL_KHR once on Android.
> * After reading some source code, it seems contentRect is complicated, so ignore it for now, `WINDOW_RESIZED` works for portrait/lamdscape mode.
> * Also check Surface.cap.currentExtent, because apparently ARM/Khronos is using that for swapchain size.
> 
> Also note that ARM best practice docs is actually in [Khronos Vulkan-Samples](https://github.com/KhronosGroup/Vulkan-Samples/tree/main/samples/performance/surface_rotation) now

I think it should be fine. There shouldn't be any danger with the window manager and the Vulkan surface disagreeing on size. The log once is a nice touch tho in case there is a bug.

**Review by chaoticbob — APPROVED** [review]
LGTM!

**Comment by hysw** [comment]
> I think it should be fine. There shouldn't be any danger with the window manager and the Vulkan surface disagreeing on size. 

It's a sanity check.
We are currently discarding resolution flag on Android, because Android window is not resizable like on desktop.
Technically Android support any swapchain resolution, but anything other than native resolution looks wrong.
